File: engine/circuits/equation_generator.py
"""backend/circuits/solver/equation_generator.py

Generates symbolic and step-by-step mathematical proofs directly from the electrical state.
Grounds educational explanations and the Bangla AI tutor in verified physics facts.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

try:
    from engine.core.parameter_resolver import format_si
except (ImportError, ValueError):
    try:
        from backend.core.parameter_resolver import format_si
    except (ImportError, ValueError):
        from core.parameter_resolver import format_si
from ..models import CircuitScene
from .mna_solver import ElectricalState


@dataclass
class DerivationStep:
    law: str
    symbolic: str
    substitution: str
    result: str
    bangla_explanation: str

    def to_dict(self) -> dict[str, str]:
        return {
            "law": self.law,
            "symbolic": self.symbolic,
            "substitution": self.substitution,
            "result": self.result,
            "bangla_explanation": self.bangla_explanation,
        }


def generate_equations(scene: CircuitScene, state: ElectricalState) -> list[dict[str, str]]:
    """Derive step-by-step educational circuit formulas from the solved state."""
    steps: list[DerivationStep] = []

    if state.status != "solved":
        return []

    # 1. Total / Equivalent Resistance Step (if single voltage source)
    v_sources = [c for c in scene.components if c.type in ("voltage_source", "battery", "cell")]
    resistors = [c for c in scene.components if c.type in ("resistor", "bulb")]

    if len(v_sources) == 1 and state.equivalent_resistance_ohm is not None:
        vs = v_sources[0]
        v_val = vs.parameters.get("voltage_v", None)
        v_num = v_val.value if v_val and v_val.value is not None else 12.0
        vs_state = state.components.get(vs.id)
        i_tot = abs(vs_state.current_a) if vs_state else (v_num / state.equivalent_resistance_ohm)

        r_eq_str = format_si(state.equivalent_resistance_ohm, "ohm")
        v_str = format_si(v_num, "V")
        i_str = format_si(i_tot, "A")

        steps.append(
            DerivationStep(
                law="equivalent_resistance",
                symbolic="R_eq = V_total / I_total",
                substitution=f"R_eq = {v_str} / {i_str}",
                result=f"R_eq = {r_eq_str}",
                bangla_explanation=f"বর্তনীর মোট তুল্য রোধ (Equivalent Resistance) হলো {r_eq_str}।",
            )
        )

        steps.append(
            DerivationStep(
                law="ohms_law_total",
                symbolic="I_total = V_total / R_eq",
                substitution=f"I_total = {v_str} / {r_eq_str}",
                result=f"I_total = {i_str}",
                bangla_explanation=f"ওহমের সূত্র অনুযায়ী বর্তনীর মূল তড়িৎপ্রবাহ I = {i_str}।",
            )
        )

    # 2. Individual Resistor Ohm's Law and Power Dissipation
    for res in resistors:
        r_state = state.components.get(res.id)
        if not r_state:
            continue
        r_param = res.parameters.get("resistance_ohm")
        r_val = r_param.value if r_param and r_param.value is not None else 100.0

        v_comp = abs(r_state.voltage_v)
        i_comp = abs(r_state.current_a)
        p_comp = r_state.power_w

        v_str = format_si(v_comp, "V")
        i_str = format_si(i_comp, "A")
        r_str = format_si(r_val, "ohm")
        p_str = format_si(p_comp, "W")

        steps.append(
            DerivationStep(
                law=f"ohms_law_{res.id}",
                symbolic=f"V_{{{res.id}}} = I_{{{res.id}}} * {res.id}",
                substitution=f"V_{{{res.id}}} = {i_str} * {r_str}",
                result=f"V_{{{res.id}}} = {v_str}",
                bangla_explanation=f"{res.id} রোধের দুই প্রান্তের বিভব পার্থক্য (Voltage Drop) হলো {v_str}।",
            )
        )

        steps.append(
            DerivationStep(
                law=f"joule_heating_{res.id}",
                symbolic=f"P_{{{res.id}}} = V_{{{res.id}}} * I_{{{res.id}}} = I^2 * R",
                substitution=f"P_{{{res.id}}} = {v_str} * {i_str}",
                result=f"P_{{{res.id}}} = {p_str}",
                bangla_explanation=f"{res.id}-এ ব্যায়িত তড়িৎ ক্ষমতা (Power Dissipation) হলো {p_str}।",
            )
        )

    # 3. Kirchhoff's Current Law (KCL) at Junctions with >= 3 connected branches
    for node in scene.nodes:
        if len(node.terminal_ids) >= 3:
            nid = node.id
            current_terms = []
            for comp in scene.components:
                c_state = state.components.get(comp.id)
                if not c_state:
                    continue
                if comp.type in ("resistor", "bulb"):
                    na = comp.terminals[0].node
                    nb = comp.terminals[1].node
                    if na == nid:
                        current_terms.append(f"-{format_si(c_state.current_a, 'A')} ({comp.id})")
                    elif nb == nid:
                        current_terms.append(f"+{format_si(c_state.current_a, 'A')} ({comp.id})")

            if current_terms:
                kcl_sub = " + ".join(current_terms).replace("+ -", "- ")
                residual = state.kcl_residuals.get(nid, 0.0)
                res_str = format_si(abs(residual), "A")
                steps.append(
                    DerivationStep(
                        law=f"kcl_{nid}",
                        symbolic=f"\\sum I_{{in}} = \\sum I_{{out}} \\implies \\sum I_{{node {nid}}} = 0",
                        substitution=f"{kcl_sub} = {res_str}",
                        result="KCL Balanced (0 A)" if abs(residual) < 1e-6 else f"Residual: {res_str}",
                        bangla_explanation=f"জাংশন {nid}-এ কার্শফের প্রথম সূত্র (KCL) রক্ষিত হয়েছে: মোট অন্তর্মুখী প্রবাহ = মোট বহির্মুখী প্রবাহ।",
                    )
                )

    return [s.to_dict() for s in steps]

File: engine/circuits/EquationGenerator.js
/**
 * circuits/EquationGenerator.js
 * Generates deterministic mathematical derivations and step-by-step physical equations
 * directly from solver facts (never hallucinated by AI).
 */

export class EquationGenerator {
  /**
   * Generate complete step-by-step derivation for the current circuit state.
   * @param {object} model - Compiled CircuitModel
   * @param {object} state - Solved ElectricalState
   * @returns {object} Structured derivation steps
   */
  static generateDerivations(model, state) {
    if (!state || !state.success) {
      return {
        summary: 'Circuit is open or unpowered. Current is 0 A.',
        steps: []
      };
    }

    const steps = [];
    const resistors = model.resistors || [];
    const vs = model.voltageSources?.[0];
    const sourceV = vs ? Number(vs.value ?? 12.0) : 12.0;
    const current = state.primaryCurrent ?? 0.0;
    const req = state.equivalentResistance ?? 0.0;

    // 1. Equivalent Resistance Derivation
    if (resistors.length >= 2 && isFinite(req) && req > 0) {
      const rValues = resistors.map(r => Number(r.value ?? 10.0));
      const rNames = resistors.map(r => r.label || r.id);
      const isSimpleSeries = model.switches.every(sw => sw.state !== 'open');

      if (isSimpleSeries) {
        steps.push({
          title: '1. সমতুল্য রোধ নির্ণয় (Equivalent Resistance in Series)',
          formula: 'R_{eq} = ' + rNames.join(' + '),
          substitution: `R_{eq} = ` + rValues.map(v => `${v} Ω`).join(' + '),
          result: `R_{eq} = ${req.toFixed(1)} Ω`,
          explanation: `অনুক্রমিক বর্তনীতে সকল রোধকের মান সরাসরি যোগ করে সমতুল্য রোধ নির্ণয় করা হয়।`
        });
      }
    }

    // 2. Total Circuit Current (Ohm's Law)
    if (sourceV > 0 && req > 0 && isFinite(req)) {
      steps.push({
        title: "2. মূল তড়িৎ প্রবাহ নির্ণয় (Total Circuit Current via Ohm's Law)",
        formula: 'I = \\frac{V}{R_{eq}}',
        substitution: `I = \\frac{${sourceV.toFixed(1)} \\text{ V}}{${req.toFixed(1)} \\, \\Omega}`,
        result: `I = ${current.toFixed(3)} A (${(current * 1000).toFixed(1)} mA)`,
        explanation: `ওহমের সূত্রানুসারে, বর্তনীর মোট বিভব পার্থক্যকে সমতুল্য রোধ দ্বারা ভাগ করলে মোট প্রবাহ পাওয়া যায়।`
      });
    }

    // 3. Voltage Drops across individual resistors
    for (const r of resistors) {
      const vDrop = state.componentVoltages?.[r.id] ?? 0.0;
      const iR = state.componentCurrents?.[r.id] ?? current;
      const rVal = Number(r.value ?? 10.0);
      const pR = state.componentPower?.[r.id] ?? (vDrop * iR);

      steps.push({
        title: `3. ${r.label || r.id}-এর প্রান্তীয় বিভব পার্থক্য ও ক্ষমতা (Potential Drop & Power)`,
        formula: `V_{${r.id}} = I \\cdot ${r.id}, \\quad P_{${r.id}} = V_{${r.id}} \\cdot I`,
        substitution: `V_{${r.id}} = ${iR.toFixed(3)} \\text{ A} \\times ${rVal.toFixed(1)} \\, \\Omega = ${vDrop.toFixed(2)} \\text{ V}`,
        result: `V = ${vDrop.toFixed(2)} V, \\quad P = ${pR.toFixed(2)} W`,
        explanation: `রোধকটির মধ্য দিয়ে প্রবাহিত তড়িৎ এবং তার রোধের গুণফলই হলো এর বিভব পতন।`
      });
    }

    // 4. Kirchhoff's Voltage Law (KVL) Verification
    if (resistors.length > 0 && sourceV > 0) {
      const vDrops = resistors.map(r => state.componentVoltages?.[r.id] ?? 0.0);
      const sumDrops = vDrops.reduce((a, b) => a + b, 0);
      const residual = Math.abs(sourceV - sumDrops);

      steps.push({
        title: "4. কার্শফের ভোল্টেজ সূত্রের প্রমাণ (KVL Closed-Loop Walk)",
        formula: '\\sum V = V_{source} - (V_{R1} + V_{R2} + \\dots) = 0',
        substitution: `${sourceV.toFixed(2)} \\text{ V} - (${vDrops.map(v => v.toFixed(2) + ' V').join(' + ')}) = ${(sourceV - sumDrops).toFixed(3)} \\text{ V}`,
        result: residual < 1e-4 ? '✓ KVL Loop Satisfied (∑V = 0)' : 'KVL Residual: ' + residual.toFixed(4) + ' V',
        explanation: `একটি আবদ্ধ লুপে সরবরাহকৃত মোট বিভব এবং সকল উপাদানের বিভব পতনের বীজগাণিতিক যোগফল শূন্য।`
      });
    }

    return {
      summary: `I = ${current.toFixed(3)} A, R_{eq} = ${req.toFixed(1)} Ω, P_{tot} = ${(state.totalPower ?? 0).toFixed(2)} W`,
      steps
    };
  }
}
